Feature add subscriber service for creating/dropping subscriptions

nathanielc opened the pull request:
Basic idea is to allow for forks or subscriptions of the data stream to be defined. A subscription is tied to database and retention policy and will send all data received to the subscription endpoint. Currently the only implementation is a UDP protocol. We do not want to create back-pressure within InfluxDB hence UDP.

Usage:

The following query creates a subscription called `sub0` that forks all incoming data to database `mydb` and policy `default` to the hosts `localhost:9090` and `remotehost:9090` in a round robin fashion. If you want to send all data to both local and remote hosts create two subscriptions.

`CREATE SUBSCRIPTION sub0 ON mydb.default HOSTS 'localhost:9090', 'remotehost:9090' TYPE 'udp'`

The following query drops the subscriptions:

`DROP SUBSCRIPTION sub0 ON mydb.default`

The following query lists all subscriptions:

`SHOW SUBSCRIPTIONS`

These subscriptions are stored in the raft meta data so the whole cluster will fork its data at each node respectively. 


TODO:

~~Need to update the QL docs and write some basic dos on how to use subcriptions.~~
~~Detect new or deleted subscriptions and start/stop them.~~

nathanielc commented on tsdb/query_executor.go:
```diff
@@ -193,6 +198,13 @@ func (q *QueryExecutor) ExecuteQuery(query *influxql.Query, database string, chu
 			case *influxql.ShowStatsStatement, *influxql.ShowDiagnosticsStatement:
 				// Send monitor-related queries to the monitor service.
 				res = q.MonitorStatementExecutor.ExecuteStatement(stmt)
+			case *influxql.CreateSubscriptionStatement, *influxql.DropSubscriptionStatement:
```
I need to trigger and event on the Subscriber server whenever a subscription is added or removed.

But I am realizing this is not the right place to tell the Subscriber to update, as only one server in a cluster will actually execute the query. Should I add a Subscriber to the MetaStore instead and have it perform this update after a save? Not entirely sure how to do this part. @pauldix

jwilder commented:
Instead of:

```
CREATE SUBSCRIPTION sub0 ON mydb.default HOSTS 'localhost:9090', 'remotehost:9090' TYPE 'udp'
```

What about something like:

```
CREATE SUBSCRIPTION sub0 ON mydb.default DESTINATIONS 'udp://localhost:9090', 'udp://remotehost:9090' 
```

I think just a `TYPE 'udp'` may not be flexible enough to add other types of endpoints in the future or have much flexibility to secure endpoints with authentication creds, etc.. 

URLs are pretty flexible for identifying the type of endpoint, it's location, specifying options, etc.. which might make it easier to expand without changing the query language each time.

For example:
```
CREATE SUBSCRIPTION sub0 ON mydb.default DESTINATIONS 'influxdb://datacenter2:8086?db=foo&u=&p=', 'riemann://remotehost: 5555', 'kafka://..., 
```

otoolep commented:
I like the URL idea.

jwilder commented:
Also, this functionality seems like it should have many of the same capabilities as the hinted handoff (durability, queue messages for failed writes, batching, back-off, node removal, etc..).  Any reason not to leverage some of that underlying infrastructure?

pauldix commented:
@jwilder the goal for this is that it would never cause InfluxDB to back up writes. We didn't want to open up the possibility of causing InfluxDB to go down if the external system did.

Although it could be done in the future. Just doesn't make sense with UDP.

pauldix commented:
I like the URL syntax that @jwilder suggested

nathanielc commented:
+1 on the URL scheme. 

Two questions:

1. How could we distinguish between multiple destinations and redundant destinations?

I might want to send data to any one of the destinations or to all of them. The use case for sending to just one is you have a backend that can scale and so you would like to balance across them,  i.e. round robin balancing.

Something like this:

`CREATE SUBSCRIPTION sub0 ON mydb.default DESTINATIONS ANY  'udp://localhost:9090', 'udp://remotehost:9090' `

or 

`CREATE SUBSCRIPTION sub0 ON mydb.default DESTINATIONS ALL 'udp://localhost:9090', 'udp://remotehost:9090', 'karka://....' `


2. As for the HH idea. We were initially thinking that this should be very lightweight and if you drop data not a big deal. Rather keep the InfluxDB server happy then apply back pressure into it. And if you wanted to do something more complex then do it outside of the DB. @pauldix Thoughts? Reasons for using HH for subscriptions vs keeping it simple?

pauldix commented:
@nathanielc I don't think the HH stuff makes sense with this initial case since it's UDP. I'd vote to keep it simple for now.

jwilder commented:
@pauldix @nathanielc For UDP transport, I agree, HH-like functionality might not be that useful since you would never know if the receiver got the write.  I was thinking more about if we _did_ need a reliable transport for a subscription.  For example, a `s3` subscription that backs up all writes to `S3 or a `http` endpoint where an application wants to ensure it always receives the post.  For those use-cases, the subscriber transport would need to durably queue the writes and retry if the receiver failed.  If would definitely need to shed load (drop new writes or expire old ones) if the queues filled up though.  Seems like those concerns could be address in that specific subscriber implementation though.

nathanielc commented:
@jwilder I like the idea of each subscriber handling the concerns of reliability.

nathanielc commented on influxql/INFLUXQL.md:
````diff
@@ -292,6 +296,22 @@ CREATE RETENTION POLICY "10m.events" ON somedb DURATION 10m REPLICATION 2;
+### CREATE SUBSCRIPTION
+
+```
+create_subscription_stmt = "CREATE SUBSCRIPTION" subscription_name "ON" db_name "." retention_policy "DESTINATIONS" ("ANY"|"ALL") host { "," host} .
````
Is this EBNF syntax correct? Not sure the `("ANY"|"ALL")` part is.

Should it be like this?
```
create_subscription_stmt = "CREATE SUBSCRIPTION" subscription_name "ON" db_name "." retention_policy "DESTINATIONS" all_any host { "," host} .

all_any = "ALL" | "ANY" 
```

otoolep commented on meta/internal/meta.proto:
```diff
@@ -210,6 +219,7 @@ message DropContinuousQueryCommand {
     required string Name = 2;
 }
 
+
```
Nit: extra blank line.

otoolep commented on services/subscriber/service_test.go:
```diff
@@ -0,0 +1,367 @@
+}
+
+func TestServiceWritePoints_ModeANY(t *testing.T) {
+
```
Style: We generally never have these leading blank lines in our functions.

otoolep commented on services/subscriber/udp.go:
```diff
@@ -0,0 +1,41 @@
+	addr string
+}
+
+func NewUDP(addr string) *UDP {
```
Can we add stats to this service? Full details on how to do this are here:

https://github.com/influxdb/influxdb/blob/master/monitor/README.md#registering-statistics-and-diagnostics

It's very easy to do.

nathanielc replied:
Will do.

otoolep commented on tsdb/query_executor.go:
```diff
@@ -46,6 +46,11 @@ type QueryExecutor struct {
 		CreateMapper(shard meta.ShardInfo, stmt influxql.Statement, chunkSize int) (Mapper, error)
 	}
 
+	// Subscriber, start/stop subscriptions
```
Comment not super clear.

nathanielc commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,204 @@
+					}
+					writers[i] = w
+					tags := map[string]string{"database": se.db, "retention_policy": se.rp, "name": se.name, "mode": si.Mode}
+					statMaps[i] = influxdb.NewStatistics(dest, "subscriber", tags)
```
@otoolep Does this look good for the stats? I didn't add it to the UDP object since here we can tag the stats with more useful information.

otoolep replied:
What does `dest` look like?

nathanielc replied:
its the URL specified in the query 'udp://host:port' should be unique to a given subscription. Does it need to be globally unique, as in all statMaps?

Looking at NewStatistics code now...

nathanielc replied:
Ok, so key is global to all statMaps, I updated it to be unique now.

otoolep replied:
The best to be *sure* you're happy with the stats is fire up the system,
generate some activity and the execute the query 'SHOW STATS' at the CLI.
How does it look?

On Fri, Oct 9, 2015 at 1:04 PM, Nathaniel Cook <notifications@github.com>
wrote:

> In services/subscriber/service.go
> <https://github.com/influxdb/influxdb/pull/4375#discussion_r41672288>:
>
> > +					return fmt.Errorf("unknown balance mode %q", si.Mode)
> > +				}
> > +				writers := make([]PointsWriter, len(si.Destinations))
> > +				statMaps := make([]*expvar.Map, len(writers))
> > +				for i, dest := range si.Destinations {
> > +					u, err := url.Parse(dest)
> > +					if err != nil {
> > +						return err
> > +					}
> > +					w, err := s.NewPointsWriter(*u)
> > +					if err != nil {
> > +						return err
> > +					}
> > +					writers[i] = w
> > +					tags := map[string]string{"database": se.db, "retention_policy": se.rp, "name": se.name, "mode": si.Mode}
> > +					statMaps[i] = influxdb.NewStatistics(dest, "subscriber", tags)
>
> Ok, so key is global to all statMaps, I updated it to be unique now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdb/influxdb/pull/4375/files#r41672288>.
>

nathanielc replied:
Looks good:

```

name: subscriber
----------------
points_written  write_failures
628             14


name: subscriber
tags: database=telegraf, destination=udp://localhost:9093, mode=ALL, name=k, retention_policy=default
points_written  write_failures
--------------  --------------
342             14
```

Thanks

otoolep replied:
Don't you feel better already? :-)

nathanielc replied:
Absolutely!
I am going to have to get monitoring setup on my side of things too soon. That was really easy and nice to use.

otoolep replied:
Remember the stats are going straight back into the _internal database, so
you can query there too, for historical insight.

On Fri, Oct 9, 2015 at 2:37 PM, Nathaniel Cook <notifications@github.com>
wrote:

> In services/subscriber/service.go
> <https://github.com/influxdb/influxdb/pull/4375#discussion_r41680738>:
>
> > +					return fmt.Errorf("unknown balance mode %q", si.Mode)
> > +				}
> > +				writers := make([]PointsWriter, len(si.Destinations))
> > +				statMaps := make([]*expvar.Map, len(writers))
> > +				for i, dest := range si.Destinations {
> > +					u, err := url.Parse(dest)
> > +					if err != nil {
> > +						return err
> > +					}
> > +					w, err := s.NewPointsWriter(*u)
> > +					if err != nil {
> > +						return err
> > +					}
> > +					writers[i] = w
> > +					tags := map[string]string{"database": se.db, "retention_policy": se.rp, "name": se.name, "mode": si.Mode}
> > +					statMaps[i] = influxdb.NewStatistics(dest, "subscriber", tags)
>
> Absolutely!
> I am going to have to get monitoring setup on my side of things too soon.
> That was really easy and nice to use.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdb/influxdb/pull/4375/files#r41680738>.
>

nathanielc commented:
This is ready from my perspective. I am able to dynamically create subscriptions and get streams of the data on my end of things. Its working nicely.

@jwilder @otoolep @pauldix How does it look now?

pauldix commented on cluster/points_writer.go:
```diff
@@ -233,6 +237,9 @@ func (w *PointsWriter) WritePoints(p *WritePointsRequest) error {
 	}
 
+	// Fire and forget points to Subscriber
+	go w.Subscriber.WritePoints(p)
```
Is there going to be bad overhead with launching  a new goroutine on every write? Just curious if this is necessary since the writer in question in UDP. I would leave it to the `Subscriber` to decide in its `WritePoints` implementation if it should be kicked off in a goroutine.

otoolep replied:
I was wondering the same thing myself, about a goroutine every write.

jwilder replied:
It might better to write this on a channel with a `default` case to prevent it from blocking on writing on the channel as well.  If this code is not async, a slow subscriber could block writes.  Something like:

```
select {
    case pointsC <- p:
        fmt.Println("sent points")
    default:
        fmt.Println("no points sent")
    }
```

otoolep replied:
Granted, the system already launches a goroutine for every write to the shards, so there is precedent. The goroutine should be amortized over the size of the points batch.

pauldix commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,216 @@
+		panic("no meta store")
+	}
+	s.Update()
+	s.ticker = time.NewTicker(time.Minute)
```
can this be configurable? Would like to see this smaller if it means that the Kapacitor user won't see data coming through for potentially a minute after creating a subscription.

Just thinking about the initial setup experience when they're going through the examples.

nathanielc replied:
Yeah, I'll make it a config option. Should it be like 5s by default so the user experience is quick the first time? Then prod systems can turn it down if they are worried.

pauldix commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,216 @@
+	for _, dbi := range dbis {
+		for _, rpi := range dbi.RetentionPolicies {
+			for _, si := range rpi.Subscriptions {
+				se := subEntry{
```
possible to pull everything in the `Subscriptions` loop into a function? Nice to avoid a bunch of code at 4 indent levels.

nathanielc replied:
done, thanks.

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,216 @@
+	return nil
+}
+
+func (s *Service) WritePoints(p *cluster.WritePointsRequest) {
```
Any chance this could be called after the service is closed?

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,216 @@
+	if s.ticker != nil {
+		s.ticker.Stop()
+	}
+	return nil
```
Any issues with not waiting for `pollForMetaUpdates` to exit? Through wait groups?

otoolep commented:
This looks quite interesting, now that I did into it.

This is a separate issue, but the polling is a little unfortunate. Perhaps this is telling us that the `Metastore` should expose a function, which returns a channel. That channel is closed any the Metastore changes. This would remove the need for polling *and* allow services such as Subscriber to respond instantaneously to Metastore changes. Would that make sense?

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,216 @@
+
+// start new and stop deleted subscriptions.
+func (s *Service) Update() error {
+	dbis, err := s.MetaStore.Databases()
```
Here it would be great instead to simply see something like:

```
for {
   <- c.MetaStore.Changed()
  .
  .
  .
}
```

nathanielc commented:
@otoolep I had a lengthy conversation with Paul about using polling. I was hoping some like what you described already existed so I could avoid polling. But since not all nodes in a cluster are part of the raft group its more difficult than just closing a channel. You would have to fire the event over a network. Unless I am mistaken on how the metastore works.

I think its something worth while but will require some effort to put together.

otoolep commented:
Ah yes, nodes are not necessarily part of the Raft group, and those that are not part of the group need to be explicitly called to hit the Raft nodes. 

I see the issue. Well, at least polling has the virtue of simplicity which makes it fairly foolproof.

jwilder commented:
@nathanielc @otoolep That functionality already exists but might not be exposed outside the metastore in a convenient way just yet.  There is https://github.com/influxdb/influxdb/blob/master/meta/store.go#L435 which is used to block a goroutine until the metastore is updated (see: https://github.com/influxdb/influxdb/blob/master/meta/rpc.go#L179).   Remote/non-raft nodes use a blocking`fetchMetaData` RPC that blocks until a meta store change occurs.  (https://github.com/influxdb/influxdb/blob/master/meta/state.go#L409). This is how meta store changes are propagated to non-raft nodes immediately.  It's similar to how Zookeeper and Etcd watches work.

I don't think there is much to do to make the remote node channel idea work.  Everything already there but might not be wired for this case.  e.g. you should be able to call `meta.WaitForDataChanged` on either a local or remote node and it will fire when the metastore is updated.  I can probably get a PR up with out too much effort.

otoolep commented:
Sounds like a great idea, assuming we can make it robust. In systems like
this we may still want some sort of polling, so we have a fail safe. The
user gets a very responsive system most of the time, but we are resilient
in the event an update is missed.

On Friday, October 9, 2015, Jason Wilder <notifications@github.com> wrote:

> @nathanielc <https://github.com/nathanielc> @otoolep
> <https://github.com/otoolep> That functionality already exists but might
> not be exposed outside the metastore in a convenient way just yet. There is
> https://github.com/influxdb/influxdb/blob/master/meta/store.go#L435 which
> is used to block a goroutine until the metastore is updated (see:
> https://github.com/influxdb/influxdb/blob/master/meta/rpc.go#L179).
> Remote/non-raft nodes use a blockingfetchMetaData RPC that blocks until a
> meta store change occurs. (
> https://github.com/influxdb/influxdb/blob/master/meta/state.go#L409).
> This is how meta store changes are propagated to non-raft nodes
> immediately. I t's simi lar to how Zookeeper and Etcd watches work.
>
> I don't think there is much to do to make the remote node channel idea
> work. Everything already there but might not be wired for this case. e.g.
> you should be able to call meta.WaitForDataChanged on either a local or
> remote node and it will fire when the metastore is updated. I can probably
> get a PR up with out too much effort.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdb/influxdb/pull/4375#issuecomment-147015340>.
>

jwilder commented:
Polling is already built in to the meta-store in addition to this functionality.

otoolep commented:
http://nathanmarz.com/blog/principles-of-software-engineering-part-1.html is an interesting read about polling *and* immediate updates.

jwilder commented:
#4427 fixes the `WaitForDataChanged` meta store func for remote nodes.  When that is merged, you should not need to poll anymore as it will release any blocked routines only when the meta store has changed.  This works for both local and remote raft nodes.

nathanielc commented:
@jwilder Great!, I'll take a look...

I just updated the PointsWriter to send data over a channel via a non blocking select statement. This should avoid the need for a new goroutine per write.

nathanielc commented:
@jwilder @otoolep @pauldix How is it looking now?

No more polling, cleaned up some style issues, no more goroutine per write(now using a channel), added stats, and config to enable/disable the service.

otoolep commented on cluster/points_writer.go:
```diff
@@ -233,6 +238,15 @@ func (w *PointsWriter) WritePoints(p *WritePointsRequest) error {
+		select {
+		case w.Subscriber.Points() <- p:
+		default:
+			w.statMap.Add(statSubWriteDrop, 1)
```
Very good, important to track this, no doubt about it.

otoolep replied:
So, perhaps track the number of requests successfully written too?

otoolep commented on meta/store_test.go:
```diff
@@ -719,6 +719,90 @@ func TestStore_DropContinuousQuery(t *testing.T) {
+	}
+}
+
+// Ensure that creating an existing continuous query returns an error.
```
Comment looks wrong.

otoolep commented on meta/store_test.go:
```diff
@@ -719,6 +719,90 @@ func TestStore_DropContinuousQuery(t *testing.T) {
+	}
+}
+
+// Ensure the store can delete a continuous query.
```
Ditto.

otoolep replied:
And further on too.

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,264 @@
+	Logger          *log.Logger
+	statMap         *expvar.Map
+	points          chan *cluster.WritePointsRequest
+	wg              *sync.WaitGroup
```
Why is this a pointer?

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,264 @@
+	points          chan *cluster.WritePointsRequest
+	wg              *sync.WaitGroup
+	closed          bool
+	mu              *sync.Mutex
```
And this?

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,264 @@
+		NewPointsWriter: newPointsWriter,
+		Logger:          log.New(os.Stderr, "[subscriber] ", log.LstdFlags),
+		statMap:         influxdb.NewStatistics("subscriber", "subscriber", nil),
+		wg:              &sync.WaitGroup{},
```
If `wg` and `mu` were not pointers, this initialization would not be needed.

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,264 @@
+	if s.MetaStore == nil {
+		panic("no meta store")
+	}
+	s.mu.Lock()
```
Why not just `defer` it?

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,264 @@
+}
+
+func (s *Service) Close() error {
+	close(s.points)
```
Ditto, why not just use `defer` for clarity?

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,264 @@
+			if err != nil {
+				lastErr = err
+				b.statMaps[i].Add(statWriteFailures, 1)
+			} else {
```
If I understand `ANY` could the code just `break` here if the mode is `ANY`? If so, you don't the `switch` statement.

nathanielc replied:
Ah yes that is a nice simplification... thanks

otoolep commented on shared/admin/js/admin.js:
```diff
@@ -309,9 +309,11 @@ var TableHeader = React.createClass({
 var TableBody = React.createClass({
     render: function() {
-        var tableRows = this.props.data.values.map(function (row) {
-            return React.createElement(TableRow, {data: row});
```
What is happening here?

nathanielc replied:
The admin console would get a js error and not display any results if a row came back without any values. Various 'show' queries do this and so I fixed it so I could use them. The fix is to check if the values array is empty and just skip drawing elements if it is.

I believe there is an issue filed for this I'll look it up and add it to the CHANGELOG of this PR.

otoolep commented on services/subscriber/service.go:
```diff
@@ -0,0 +1,258 @@
+	var lastErr error
+	for range b.writers {
+		i := b.i
+		w := b.writers[i]
```
What is the point of the modulo stuff?

nathanielc replied:
The module code rotates through destinations, otherwise each write would always try the first destination first and it would get a disproportionate amount of data.

otoolep commented:
LGTM, one question.

jwilder commented on cmd/influxd/run/config_test.go:
```diff
@@ -45,6 +45,9 @@ bind-address = ":4444"
 enabled = true
 
+[subscriber]
+enabled = false
```
Should probably use `true` since the default for a bool is `false`.

jwilder commented on services/subscriber/udp.go:
```diff
@@ -0,0 +1,40 @@
+func (u *UDP) WritePoints(p *cluster.WritePointsRequest) (err error) {
+	var addr *net.UDPAddr
+	var con *net.UDPConn
+	addr, err = net.ResolveUDPAddr("udp", u.addr)
```
Not sure if `UDP` instances are long-lived or not, but resolving the addr and creating the `con` below could probably be done once at some point.  Not a big deal though.

jwilder commented:
LGTM. :+1:

manishjain002 commented:
I have a use case, please see if this make sense in implementing in "Subscribe".
I have an application for monitoring of my sensors data that's getting logged. Before "Subscribe" , I had only one way, i.e. Pull the data from the DB server using "polling". But with subscription, I get data Pushed.
But there is a problem, I am getting all the data that is getting stored in the DB, I would be interested in filtering data. What I want is "selective Subscription".

beckettsean commented:
@manishjain002 what you are describing is a feature request for Kapacitor. The repo for that project will be public as of tomorrow (Dec 1st) so I encourage you to [open an issue](https://github.com/influxdb/kapacitor/issues/new) there.
